doc/design/bucketpath.rst

```diff
@@ -0,0 +1,420 @@
+
+   The subsequent subsections include code snippets intended primarily for the implementer's reference.
+   Therefore, comments and docstrings within the sudo code may need adaptation for the actual implementation.
+
+
+Pathlike
+--------
+
+.. code-block:: python
+
+    from typing import Protocol
+
+    class Pathlike(Protocol):
+
+        @property
+        def name:
+            """
+            A string representing the final path component, excluding the drive and root, if any.
+            """
+
+        @property
+        def suffix:
+            """
+            The file extension of the final component, if any.
+            """
+
+        @property
+        def root:
+            """
+            A string representing the root, if any.
+            """
+
+        @property
+        def parent:
+            """
+            The logical parent of this path.
+            """
+
+        def as_uri():
+            """
+            Represent the path as a file URI. Can be used to reconstruct the location/path.
+            """
+
+        def exists():
+            """
+            Return True if the path points to an existing file or directory.
+            """
+
+        def is_dir():
+            """
+            Return True if the path points to a directory, False if it points to another kind of file.
+            """
+
+        def is_file():
+            """
+            Return True if the path points to a regular file, False if it points to another kind of file.
+            """
+
+        def read(chunk_size: int = 8192) -> Iterable[ByteString]:
+            """
+            TBD
+            """
+
+        def write(data: ByteString | BinaryIO | Iterable[ByteString]):
+            """
+            TBD
+            """
+
+        def rm():
+            """
+            If exists() and is_file yields true for this path, the path will be deleted,
+            otherwise exception will be thrown.
+            """
+
+        def rmdir(recursive: bool = False):
+            """
+            TBD
+            """
+
+        def joinpath(*pathsegements)
+            """
+            Calling this method is equivalent to combining the path with each of the given pathsegments in turn
+            """
+
+        def walk():
+            """
+            TBD
+            """
+
+        def iterdir():
+            """
+            TBD
+            """
+
+        # Overload / for joining, see also joinpath or `pathlib.Path`.
+        def __truediv__():
+            """
+            TBD
+            """
+
+
+Concrete Paths (Backends)
+-------------------------
+
+Implement the ``Pathlike`` protocol for each specific backend. Concrete paths must include:
+
+- A ``protocol`` member indicating the associated protocol.
+- Backend-specific methods for creation.
+- Implementations for all methods and properties required by the ``Pathlike`` protocol
+
+Each backend must implement the ``as_uri`` method in a way that the location is clearly identifiable.
```

Comment:
I am not sure whether this can be provided in all cases. 
It might depend on the definition of "uri".
What is an URI in this context?

Reply:
If a URI is understood as a string identifying a bucket or path in a bucket from outside, then some parts of the URI might be unknown to a UDF.

Reply:
For details on this please see the example at the bottom (factory section).